consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+
+	// GetSnapshot returns copy of the current snapshot
+	GetSnapshot() (*ProposerSnapshot, bool)
+}
+
+// check if interface is matching implementation
+var _ ProposerCalculator = (*proposerCalculator)(nil)
+
+type proposerCalculator struct {
+	// current snapshot
+	snapshot *ProposerSnapshot
+
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculator(config *runtimeConfig, logger hclog.Logger) (*proposerCalculator, error) {
+	snap, err := NewProposerSnapshotFromState(config, logger)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &proposerCalculator{
+		snapshot: snap,
+		logger:   logger,
+	}, nil
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculatorFromSnapshot(pcs *ProposerSnapshot, logger hclog.Logger) *proposerCalculator {
+	return &proposerCalculator{
+		snapshot: pcs.Copy(),
+		logger:   logger,
+	}
+}
+
+// Get copy of the proposers' snapshot
+func (pc *proposerCalculator) GetSnapshot() (*ProposerSnapshot, bool) {
+	if pc.snapshot == nil {
+		return nil, false
+	}
+
+	return pc.snapshot.Copy(), true
+}
+
+// GetLatestProposer returns latest calculated proposer if any
+func (pc *proposerCalculator) GetLatestProposer(round, height uint64) (types.Address, bool) {
+	// round must be same as saved one and proposer must exist
+	if pc.snapshot == nil || pc.snapshot.Proposer == nil || pc.snapshot.Round != round || pc.snapshot.Height != height {
+		pc.logger.Info("Get latest proposer not found", "height", height, "round", round,
+			"pc height", pc.snapshot.Height, "pc round", pc.snapshot.Round)
+
+		return types.ZeroAddress, false
+	}
+
+	pc.logger.Info("Get latest proposer",
+		"height", height, "round", round, "address", pc.snapshot.Proposer.Metadata.Address)
+
+	return pc.snapshot.Proposer.Metadata.Address, true
+}
+
+// CalcProposer calculates next proposer
+func (pc *proposerCalculator) CalcProposer(round, height uint64) (types.Address, error) {
+	if height != pc.snapshot.Height {
+		return types.ZeroAddress, fmt.Errorf("invalid height - expected %d, got %d", pc.snapshot.Height, height)
+	}
+
+	// optimization -> return current proposer if already calculated for this round
+	if pc.snapshot.Round == round && pc.snapshot.Proposer != nil {
+		return pc.snapshot.Proposer.Metadata.Address, nil
+	}
+
+	// do not change priorities on original snapshot while executing CalcProposer
+	// if round = 0 then we need one iteration
+	proposer, err := incrementProposerPriorityNTimes(pc.snapshot.Copy(), round+1)
+	if err != nil {
+		return types.ZeroAddress, err
+	}
+
+	pc.snapshot.Proposer = proposer
+	pc.snapshot.Round = round
+
+	pc.logger.Info("New proposer calculated", "height", height, "round", round, "address", proposer.Metadata.Address)
+
+	return proposer.Metadata.Address, nil
+}
+
+func (pc *proposerCalculator) Update(blockNumber uint64, config *runtimeConfig, state *State) error {
```

Comment:
Move the two static parameters `config` and `state` to the constructor of the `proposerCalculator` instead.

Reply:
We want to have `config` and `state` as parameters of `proposerCalculator`?

Reply:
Why not, they are static parameters. Those parameters are not related to the `Update` action.

Reply:
Fixed https://github.com/0xPolygon/polygon-edge/pull/1018/commits/f0879b3c9bf91004bf3dd6f1243cc2be6cf960d5